### PR TITLE
Write support

### PR DESCRIPTION
**Pull request by s3bk**
Remaining questions:
 
## How to deal with objects from multiple files? ##
Right now indirect references only store object and generation number. This is sufficient to retrieve them from the correct file.
However with writing one may end up having more than one file open at one time (1+ input, 1 output for example).
This lends itself to easy user errors by plugging the wrong reference into the wrong file.

Possible solutions:
 - ignore it
 - Keep a reference to the `Rc<Storage>` in `Ref` (pollutes everything with `B: Backend`)
 - Keep a `Rc<dyn Resolve + Update>` in `Ref`.
 - generate a File/Storage ID and keep that.

### CODE REVIEW

**Comment by s3bk** [comment]
@Ploppz what do you think?
@divergentdave could you take a look as well? I am probably breaking a few things, especially in regards to Ref/Rc fields.

File: pdf/src/backend.rs [hunk 1]
```diff
@@ -111,18 +111,17 @@ impl<T> Backend for T where T: Deref<Target=[u8]> { //+ DerefMut<Target=[u8]> {
...
     /// End index (exclusive)
-    fn end(&self) -> Option<usize> { None }
+    fn end(&self) -> Option<usize>;
 
     /// `len`: the size of whatever container that is being indexed
     fn to_range(&self, len: usize) -> Result<Range<usize>> {
         match (self.start(), self.end()) {
             (None, None) => Ok(0 .. len),
-            (Some(start), _) if start <= len => Ok(start .. len),
+            (Some(start), None) if start <= len => Ok(start .. len),
```
Comment (s3bk) [inline pdf/src/backend.rs:124]: Actually a bugfix.

File: pdf/src/enc.rs [hunk 2]
```diff
@@ -200,16 +299,59 @@ fn dct_decode(data: &[u8], _params: &DCTDecodeParams) -> Result<Vec<u8>> {
     Ok(pixels)
 }
 
+fn lzw_decode(mut data: &[u8], params: &LZWFlateParams) -> Result<Vec<u8>> {
```
Comment (s3bk) [inline pdf/src/enc.rs:302]: LZW might be broken.

**Comment by s3bk** [comment]
It looks like the page tree is slightly broken right now.

**Comment by Ploppz** [comment]
Skimmed through it and the write implementation looks good from a glance!

- Is there a benefit to keeping Object and ObjectWrite traits separate or could ObjectWrite function be put inside Object?
(I like that you replaced `Object::serialize()` with a `to_primitive()` function)
- There seems to be a lot of changes not related to write functionality?
- Not sure how to answer your question, sorry. Maybe you can wait with solving that.

**Comment by s3bk** [comment]
> Is there a benefit to keeping Object and ObjectWrite traits separate or could ObjectWrite function be put inside Object?
It means we can delay implementing write support for all types.

> There seems to be a lot of changes not related to write functionality?
Yes, I fixed a bunch of things I ran into.

**Comment by s3bk** [comment]
It builds again and tests pass. I think a merge is possible.

**Pull request merged**